semanticdb3/semanticdb3.md

```diff
@@ -0,0 +1,491 @@
+
+### Range
+
+```protobuf
+message Range {
+  int32 start_line = 1;
+  int32 start_character = 2;
+  int32 end_line = 3;
+  int32 end_character = 4;
+}
+```
+
+`Range` in SemanticDB directly corresponds to `Range` in LSP [\[2\]][2].
+It represents a range between start and end points in a document. Both points
+are represented by zero-based line and zero-based character offsets.
+The start point is inclusive, while the end point is exclusive.
+
+### Symbol
+
+Symbols are tokens that are used to correlate identifiers and definitions.
+In the SemanticDB model, symbols are represented as strings.
+At the moment, the symbol format is defined by the needs of the Scala implementations.
+In the future, we are planning to pay more attention to other languages,
+and we may decide to expand the format to accommodate namespacing rules not present in Scala.
+
+**Global symbols**. Correspond to a definition that can be referenced outside
+the compilation unit where the definition is defined.
+Global symbol format is a concatenation of signatures of the owner chain
+of the corresponding global definition, where:
+  * The owner chain of a definition is a list of its enclosing definitions
+    starting with the outermost one, with the outermost definition being
+    either `_empty_` (the special empty package [\[17\]][17]) or
+    `_root_` (the special root package [\[18\]][18]). For example,
+    for the standard `Int` class, the owner chain is `[_root_, scala, Int]`.
+  * The signature of a definition is:
+    * For a method, concatenation of its name, its JVM method descriptor [\[19\]][19]
+      and a dot (`.`). The JVM method descriptor is used to distinguish
+      overloaded methods as mandated by the Scala Language Specification [20].
+      For any other term definition (package, object, val or var),
+      concatenation of its name and a dot (`.`).
+    * For a type definition (class, type alias or type member),
+      concatentation of its name and a pound sign (`#`).
+
+For example, the standard `Int` class must be modelled by a global symbol `_root_.scala.Int#`.
+
+**Local symbols**. Correspond to a definition that isn't global (see above).
+Local symbol format is deliberately unspecified except for two restrictions:
+  * Local symbols must start with `local`, so that they can be easily
+    distinguished from global symbols.
+  * Local symbols must be unique within the underlying document.
+
+For example, `x` in `def identity[T](x: T): T` may be modelled by local symbols
+`local0`, `local_x`, `local_identity_x`, as long as these names are unique within
+the underlying document. The same logic applies to the type parameter `T`,
+which is also a local definition.
+
+**Placeholder symbols**. Are used to model original snippets of code in [Synthetics](#synthetic).
+Must not be used outside `Synthetic.text` documents. Placeholder symbols are
+always equal to an asterisk (`*`).
+
+### Identifier
+
+```protobuf
+message Identifier {
```

Comment:
One issue I have with `Identifier` is that it doesn't specify **what** it's identifying. In LSP there is `TextDocumentIdentifier`, BSP has `BuildTargetIdentifier` while here we have an unqualified `Identifier`. What about calling this `SymbolIdentifier` instead? That would make the relationship between `SymbolIdentifier` and `SymbolInformation` clearer IMO.

Reply:
After more thinking I don't like `SymbolIdentifier` since that intuitively means "the ID for this symbol". `SymbolOccurrence` is the most accurate and self-explanatory name I can think of, and it would match with clangs IndexWhileBuilding usage of "Occurrence".

Reply:
That's a valid argument. In PL, `Identifier` is unambiguous, but within a larger context this name may probably be not so clear. Let's think about alternatives.

What I like about `Identifier` is that it promotes a text-centric view of the program, where we aren't talking compiler internals, but are appealing to what is written in the source file. In that sense, I liked `Identifier` and `Definition`, because that's what's written in source code. 

`SymbolOccurrence` and `SymbolInformation` are pushing us away from that model. For example, with a text-centric view, you can have an identifier or a definition without any attached symbol, e.g. created by a parser. However, if you call them symbol occurrence or information, then it kinda doesn't make sense to have these data structures without semantic information. Now, this is not something that we have been doing with SemanticDB yet, but this illustrates the idea.

Reply:
`SymbolOccurrence` seems to be a thing in swift-clang btw, https://github.com/apple/swift-clang/blob/stable/include/clang/Tooling/Refactor/SymbolOccurrenceFinder.h

My favorite remains `SymbolOccurrence`, and I think it's the most consistent name to pair up with `SymbolInformation`

Reply:
Let me think about this over the weekend.